### PR TITLE
Implement horizontal input fusion.

### PR DESCRIPTION
**Pull request by trentlo**
Extend horizontal fusion to support fusion of reduction instructions.

The PR for the code generation of parallel reduction is [here](https://github.com/tensorflow/tensorflow/pull/42717). Potential performance gain can also be inferred by the numbers listed in the code gen PR.

### CODE REVIEW

**Comment by trentlo** [comment]
@thomasjoerg could you help to review this PR when you has a moment? Thanks!

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 1]
```diff
@@ -506,5 +504,24 @@ HloInstruction::FusionKind ChooseFusionKind(const HloInstruction& /*producer*/,
...
+  return absl::c_all_of(instr.users(), [&](const HloInstruction* user) {
+    if (user->opcode() == HloOpcode::kGetTupleElement) {
+      // Skip GTE.
+      return IsConsumerTheOnlyNonRootUser(*user, consumer);
+    } else if (user == &consumer) {
+      // `user` is `consumer`.
+      return true;
+    } else if (user == user->parent()->root_instruction()) {
+      // Consumed by ROOT is always fine, since it is impossible to create
+      // cycles through ROOT.
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/gpu_fusible.cc:518]: Not clear why we care about cycles here. Please clarify the comment.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include "tensorflow/core/platform/errors.h"
+
+namespace xla {
+namespace gpu {
+
+namespace {
+
+// Gets the representative input shape of the multi-output fusion.
+Shape GetInputShapeForMultiOutputFusion(const HloInstruction& instr) {
+  // Get the major node used in the emitter.
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc:36]: Get the HLO that determines the what emitter will be used.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "tensorflow/compiler/xla/service/hlo_pass_interface.h"
+#include "tensorflow/core/platform/macros.h"
+
+namespace xla {
+namespace gpu {
+
+// This optimization pass horizontally fuses kInput fusions to both reduce the
+// kernel launch overhead and increase parallelism degree. See
+// GpuHorizontalFusion for general description and motivation about horizontal
+// fusion. GpuHorizontalFusion deals with kLoop fusions while this pass deals
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.h:31]: Would it make sense to s/GpuHorizontalFusion/GpuHorizontalLoopFusion? Can be a separate PR though.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    size_t fusion_anchor_id = 0;
+    for (size_t j = 1; j < candidates.size(); ++j) {
+      HloInstruction* fusion_anchor = candidates[fusion_anchor_id];
+      HloInstruction* fused = candidates[j];
+      if (ShapesCompatibleForMultiOutputFusion(*fusion_anchor, *fused) &&
+          !FusionWouldBeTooLarge(*fusion_anchor, *fused)) {
+        VLOG(3) << absl::StrCat("Fuse ", fused->ToString(), " into ",
+                                fusion_anchor->ToString());
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc:122]: `VLOG(3) << "Fuse " << fused->ToString() << " into " << fusion_anchor->ToString();`
Same below.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+  std::sort(fusion_instrs.begin(), fusion_instrs.end(),
+            [&](const HloInstruction* a, const HloInstruction* b) {
+              Shape shape_a = GetInputShapeForMultiOutputFusion(*a);
+              Shape shape_b = GetInputShapeForMultiOutputFusion(*b);
+              if (shape_a.rank() != shape_b.rank()) {
+                return shape_a.rank() < shape_b.rank();
+              } else if (ShapeUtil::ElementsIn(shape_a) !=
+                         ShapeUtil::ElementsIn(shape_b)) {
+                // Sort according to element size so that roughly the same input
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc:86]: This would order shapes [128,256] and [256,128] in an arbitrary order, right? If so, we may miss out on fusion opportunities because two equal shapes may not be sorted next to each other.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion_test.cc [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        output_shape, mul, const0, {1}, reduce_computation));
+    var_outs.push_back(reduce);
+  }
+  builder.AddInstruction(HloInstruction::CreateTuple(var_outs));
+  module->AddEntryComputation(builder.Build());
+
+  // Verify that horizontal fusion is kicked in. Check that there are multiple
+  // `reduce` instructions fused into the same fusion. 6 is just a randomly
+  // picked number as we don't exactly know how large the fusion will be
+  // created.
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/horizontal_input_fusion_test.cc:135]: Why don't we know? Is it because the "fusion too large" heuristic will kick in at some time?

**Review by thomasjoerg — DISMISSED** [review]

File: tensorflow/compiler/xla/service/gpu/BUILD [hunk 7]
```diff
@@ -1769,6 +1771,47 @@ tf_cc_test(
...
+        "//tensorflow/core:lib_internal",
+        "@com_google_absl//absl/container:flat_hash_set",
+    ],
+)
+
+tf_cc_test(
+    name = "horizontal_input_fusion_test",
+    srcs = ["horizontal_input_fusion_test.cc"],
+    deps = [
+        ":fusion_merger",
```
[the remark this answers is not in the document]
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/BUILD]: My oversight. Cleaned them up.

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 8]
```diff
@@ -143,29 +143,27 @@ bool IsInputFusibleReduction(const HloInstruction& instr) {
...
-    if (!instr->IsMultiOutputFusion()) {
-      return fused_expression_root;
-    }
-    // If possible, we want to pick a reduction-to-vector operand of the
-    // fusion root, because it has the most constraints.
-    for (const auto* inst : fused_expression_root->operands()) {
-      if (IsReductionFromOrToContiguousDimensions(*inst)) {
-        return inst;
-      }
+const HloInstruction* GetMajorNodeForMultiOutputFusion(
```
[the remark this answers is not in the document]
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Revised according to your suggestion. Thanks!

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 9]
```diff
@@ -506,5 +504,24 @@ HloInstruction::FusionKind ChooseFusionKind(const HloInstruction& /*producer*/,
                                   : HloInstruction::FusionKind::kLoop;
 }
 
+bool IsConsumerTheOnlyNonRootUser(const HloInstruction& instr,
+                                  const HloInstruction& consumer) {
+  return absl::c_all_of(instr.users(), [&](const HloInstruction* user) {
+    if (user->opcode() == HloOpcode::kGetTupleElement) {
+      // Skip GTE.
+      return IsConsumerTheOnlyNonRootUser(*user, consumer);
+    } else if (user == &consumer) {
```
[the remark this answers is not in the document]
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Revised.

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc (hunk 1 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Revised. We simply should not have these comments here. They are confusing.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc (hunk 2 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc]: Done.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc (hunk 5 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc]: I revised the comparison function.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc (hunk 4 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.cc]: Changed.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.h (hunk 3 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_input_fusion.h:31]: That makes sense. I will make another PR to do this.

File: tensorflow/compiler/xla/service/gpu/horizontal_input_fusion_test.cc (hunk 6 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_input_fusion_test.cc]: Right. It is because of the FusionWouldBeTooLarge constraint. I added the reason into the comments.

**Review by trentlo — COMMENTED** [review]
Thank you for the review! @thomasjoerg 

I should have addressed all the comments. Please help to take a look again.

**Comment by trentlo** [comment]
@thomasjoerg ping~

**Review by thomasjoerg — APPROVED** [review]
